resolver/src/main/java/io/netty/resolver/HostsFileEntriesProvider.java

```diff
@@ -0,0 +1,311 @@
+         * are not provided, the system default {@link Charset} is used for decoding
+         * @return a new {@link HostsFileEntriesProvider}
+         * @throws IOException file could not be read
+         */
+        HostsFileEntriesProvider parse(File file, Charset... charsets) throws IOException;
+
+        /**
+         * Performs the parsing operation using the provided reader of hosts file format.
+         *
+         * @param reader the reader of hosts file format
+         * @return a new {@link HostsFileEntriesProvider}
+         */
+        HostsFileEntriesProvider parse(Reader reader) throws IOException;
+
+        /**
+         * Parses the hosts file at standard OS location using the system default {@link Charset} for decoding.
+         *
+         * @return a new {@link HostsFileEntriesProvider}
+         */
+        HostsFileEntriesProvider parseSilently();
+
+        /**
+         * Parses the hosts file at standard OS location using the given {@link Charset}s one after another until
+         * parse something or none is left.
+         *
+         * @param charsets the {@link Charset}s to try as file encodings when parsing
+         * @return a new {@link HostsFileEntriesProvider}
+         */
+        HostsFileEntriesProvider parseSilently(Charset... charsets);
+
+        /**
+         * Parses the provided hosts file using the given {@link Charset}s one after another until
+         * parse something or none is left. In case {@link Charset}s are not provided,
+         * the system default {@link Charset} is used for decoding.
+         *
+         * @param file the file to be parsed
+         * @param charsets the {@link Charset}s to try as file encodings when parsing, in case {@link Charset}s
+         * are not provided, the system default {@link Charset} is used for decoding
+         * @return a new {@link HostsFileEntriesProvider}
+         */
+        HostsFileEntriesProvider parseSilently(File file, Charset... charsets);
+    }
+
+    /**
+     * Creates a parser for {@link HostsFileEntriesProvider}.
+     *
+     * @return a new {@link HostsFileEntriesProvider.Parser}
+     */
+    public static Parser parser() {
+        return new ParserImpl();
+    }
+
+    static final HostsFileEntriesProvider EMPTY =
+            new HostsFileEntriesProvider(
+                    Collections.<String, List<InetAddress>>emptyMap(),
+                    Collections.<String, List<InetAddress>>emptyMap());
+
+    private final Map<String, List<InetAddress>> ipv4Entries;
+    private final Map<String, List<InetAddress>> ipv6Entries;
+
+    HostsFileEntriesProvider(Map<String, List<InetAddress>> ipv4Entries, Map<String, List<InetAddress>> ipv6Entries) {
+        this.ipv4Entries = Collections.unmodifiableMap(new HashMap<String, List<InetAddress>>(ipv4Entries));
+        this.ipv6Entries = Collections.unmodifiableMap(new HashMap<String, List<InetAddress>>(ipv6Entries));
```

Comment:
@violetagg question... should we use a `LinkedHashMap` here or doesn't it matter ?

Reply:
I don't think that we have a use case for `LinkedHashMap` where we will need the insertion order. We always need the info for one particular `hostname`